nixos/modules/services/monitoring/loki.nix

```diff
@@ -0,0 +1,108 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) types mkIf concatStringsSep mkOption mkEnableOption;
+
+  cfg = config.services.loki;
+
+  prettyJSON = conf:
+    pkgs.runCommand "loki-config.json" { } ''
+      echo '${builtins.toJSON conf}' | ${pkgs.jq}/bin/jq 'del(._module)' > $out
+    '';
+
+in {
+  options.services.loki = {
+    enable = mkEnableOption "loki";
+
+    user = mkOption {
+      type = types.str;
+      default = "loki";
+      description = ''
+        User under which the Loki service runs.
+      '';
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "loki";
+      description = ''
+        Group under which the Loki service runs.
+      '';
+    };
+
+    dataDir = mkOption {
+      type = types.path;
+      default = "/var/lib/loki";
+      description = ''
+        Specify the directory for Loki.
+      '';
+    };
+
+    configuration = mkOption {
+      type = types.attrs;
+      default = {};
+      description = ''
+        Specify the configuration for Loki in Nix.
+        Note that if specified, <option>configFile</option> takes precedence over <option>configuration</option>.
+      '';
+    };
+
+    configFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = ''
+        Specify a configuration file that Loki should use.
+      '';
+    };
+
+    extraFlags = mkOption {
+      type = types.listOf types.str;
+      default = [];
+      description = ''
+        Specify additional command line flags that are passed to Loki.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [{
+      assertion = (cfg.configuration == {} -> cfg.configFile != null);
+      message  = ''
+        Please specify either
+        'services.loki.configuration' or
+        'services.loki.configFile'.
+      '';
+    }];
+
+    users.groups.${cfg.group} = { };
+    users.users.${cfg.user} = {
+      description = "Loki Service User";
+      group = cfg.group;
+      home = cfg.dataDir;
+      createHome = true;
+      isSystemUser = true;
+    };
+
+    systemd.services.loki = {
+      description = "Loki Service Daemon";
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = let
+        conf = if cfg.configFile == null
+               then prettyJSON cfg.configuration
+               else cfg.configFile;
+      in
+      {
+        ExecStart = "${pkgs.grafana-loki}/bin/loki --config.file ${conf} ${concatStringsSep " " cfg.extraFlags}";
```

Comment:
```suggestion
        ExecStart = "${pkgs.grafana-loki}/bin/loki --config.file ${conf} ${lib.escapeShellArgs cfg.extraFlags}";
```

Reply:
In addition to your suggested change, I updated the option `extraFlags` to include a hint about the escaping, because this is, even though it would seem to make sense to use this more often, quite unusual in nixpkgs:

```diff
@@ -58,8 +59,10 @@ in {
     extraFlags = mkOption {
       type = types.listOf types.str;
       default = [];
+      example = literalExample [ "--server.http-listen-port=3101" ];
       description = ''
-        Specify additional command line flags that are passed to Loki.
+        Specify a list of additional command line flags,
+        which get escaped and are then passed to Loki.
       '';
     };
   };
```

The semantics are a little different now, because with `concatStringsSep` you could for example specify a custom flag like
```
extraFlags = [ "-some-flag value" ];
```
Which would just work as expected. If you specify it the same way and run it through `escapeShellArgs`, the parameter will be passed as `'-some-flag value'`. This breaks the previously known behaviour and throws the following error:
```
flag provided but not defined: -some-flag value
```

I think with the added hint and example this should be fine, thank's for the suggestion :+1:

Reply:
So if you use `extraFlags = [ '-some-flag' 'value' ];` it works?

Reply:
Yes, with double-quotes that works, also would `extraFlags = [ "-some-flag=value" ];`. :+1: